proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,259 @@
+}
+```
+
+The `EpochRewardHistory` is updated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch
+rewards and vote account rewards become available at this time. Once the vote
+account rewards have been distributed, a `RewardHistoryEntry` will be added to
+the `EpochRewardHistory` array with the total amount of reward, a distributed
+value equal to the sum of the vote account rewards. If the length of the history
+array exceeds the maximum depth (512), the oldest entries will be removed from
+the array (i.e., "popped").
+
+The `root_hash` is the hash of all the `EpochRewardReserve` accounts hashes.
+Initially, at the start of the epoch's first block, the `root_hash` is set to
+None. Once all the `EpochRewardReserve` accounts are filled and the reward
+computation phase is complete, the `root_hash` is calculated by hashing all the
+`EpochRewardReserve` hashes.
+
+```
+root_hash = Hash(EpochRewardReserve[..].reserve_hash)
+
+
+
+                     +-------------+
+                     |             |
+                     |  root_hash  |
+                     |             |
+                     +------+------+
+                            |
+    +--------------+--------+-----------+
+    |              |                    |
++---+----+   +-----+--+             +---+-----+
+|reserve |   | reserve|   ......    | reserve |
+| hash   |   |  hash  |             |  hash   |
++--------+   +--------+             +---------+
+
+```
+
+### `EpochRewardReserve` Sysvar Account
+
+The address for a particular `EpochRewardReserve` account is derived from the
+base Sysvar RewardReserve address, by hashing
+`SysvarRewardReserve11111111111111111111111111111111` with the u64 block height
+for the corresponding reward distribution block during the reward
+distribution phase.
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve11111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   reserve_hash: Hash,
+}
+```
+
+The account balance of `RewardsReserve` is equal to the total amount (in
+lamports) of all rewards to be distributed in a block. The `reserve_hash` is
+computed from the rewards in the block as follows.
+
+```
+reserve_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, M `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick time without much performance impact for the overall block
+time. Given that each block time is 64 ticks time long, a total of 4,096
+accounts can be distributed per block. The total number of blocks needed tto
+distributed rewards is
```

Comment:
This chunking happens in the rewards calculation phase, right? I feel like this information makes more sense in that section.